proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+# Move Function + "Use After Move" Diagnostic
+
+* Proposal: [SE-NNNN](nnnn-move-function.md)
+* Authors: [Michael Gottesman](https://github.com/gottesmm), [Andrew Trick](https://github.com/atrick)
+* Review Manager: TBD
+* Status: Implemented on main as stdlib SPI (`_move` instead of `move`)
+
+<!--
+*During the review process, add the following fields as needed:*
+
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN) or [apple/swift-evolution-staging#NNNNN](https://github.com/apple/swift-evolution-staging/pull/NNNNN)
+* Decision Notes: [Rationale](https://forums.swift.org/), [Additional Commentary](https://forums.swift.org/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+-->
+
+## Introduction
+
+In this document, we propose adding a new function called `move` to the swift
+standard library, which ends the lifetime of a specific local `let`,
+local `var`, or `consuming` parameter. In order to enforce this, the compiler will
+emit a flow-sensitive diagnostic upon any uses that are after the move
+function. As an example:
+
+```
+useX(x) // do some stuff with local variable x
+
+// Ends lifetime of x, y's lifetime begins.
+let y = move(x) // [1]
+
+useY(y) // do some stuff with local variable y
+useX(x) // error, x's lifetime was ended at [1]
+
+// Ends lifetime of y. Since _ is no-op, we perform an actual release here.
+let _ = move(y) // [2]
+useX(x) // error, x's lifetime was ended at [1]
+useY(y) // error, y's lifetime was ended at [2]
+```
+
+This allows the user to influence the uniqueness of COW data structures
```

Comment:
I would start the "Motivation" section here, and reword things some to focus
on the problems that you're trying to solve:
```
Performance-critical code can suffer from unexpected copy-on-write operations
and unnecessary retain/release calls.
Consider the following array/uniqueness example:
```